# Solution
***

## a)

In [1]:
#Imports
import numpy as np
import pandas as pd
import matplotlib.pyplot as plt
import statsmodels.formula.api as smf

In [2]:
#Creating data
x = np.array([1, 2, 3, 4, 5])
y = np.array([0.83, 6.99, 7.57, 5.99, 9.84])
n = x.size

In [3]:
#Calculating a1
frac1 = n * (x * y).sum() - y.sum() * x.sum()
frac2 = n * (x**2).sum() - x.sum()**2
a1 = frac1 / frac2

In [4]:
#Calculating a0
a0 = y.mean() - a1 * x.mean()

In [5]:
#Function
f = a0 + a1 * x
print(f)

[2.84  4.542 6.244 7.946 9.648]


## b)

In [6]:
df = pd.DataFrame({'x' : x, 'y' : y})
model1 = smf.ols('y ~ x', data = df)
m1 = model1.fit()
print(m1.params)

df_new = pd.DataFrame({'x' : [3], 'y' : [0]})
pred1 = m1.predict(df_new)
print(f'{pred1[0]:.3f}')

Intercept    1.138
x            1.702
dtype: float64
6.244


## c)

In [7]:
#Prediction for different K
x0 = 3
distances = np.abs(x - x0)
print(distances)

[2 1 0 1 2]


In [8]:
#Sort distances by index
sorted_idx = np.argsort(distances)
print(sorted_idx)

[2 1 3 0 4]


In [9]:
#K = 1
k1 = np.mean(y[sorted_idx[:1]])
print(k1)

7.57


In [10]:
#K = 3
k3 = np.mean(y[sorted_idx[:3]])
print(round(k3, 2))

6.85


In [11]:
#K = 5
k5 = np.mean(y[sorted_idx[:5]])
print(round(k5, 2))

6.24


In [12]:
#Print results
print(f'Predicted value for K = 1 is {k1:.2f}')
print(f'Predicted value for K = 3 is {k3:.2f}')
print(f'Predicted value for K = 5 is {k5:.2f}')

Predicted value for K = 1 is 7.57
Predicted value for K = 3 is 6.85
Predicted value for K = 5 is 6.24


## d)

In [13]:
#Function for KNN
def knn(x0, x, y, K = 20):
    d = np.abs(x - x0)
    o = np.argsort(d)[:K]
    ypred = np.mean(y[o])
    return ypred

In [14]:
#Experimenting with the function
K = 3
x0 = 3
d = np.abs(x - x0)
o = np.argsort(d)[:K]
ypred = np.mean(y[o])
print(round(ypred, 2))

6.85


# Task 2
***

## a)
***

In [15]:
#Loading data
from ISLP import load_data

College = load_data('College')
College.describe(include = 'all')

,Private,Apps,Accept,Enroll,Top10perc,Top25perc,F.Undergrad,P.Undergrad,Outstate,Room.Board,Books,Personal,PhD,Terminal,S.F.Ratio,perc.alumni,Expend,Grad.Rate
count,777,777.000000,777.000000,777.000000,777.000000,777.000000,777.000000,777.000000,777.000000,777.000000,777.000000,777.000000,777.000000,777.000000,777.000000,777.000000,777.000000,777.00000
unique,2,NaN,NaN,NaN,NaN,NaN,NaN,NaN,NaN,NaN,NaN,NaN,NaN,NaN,NaN,NaN,NaN,NaN
top,Yes,NaN,NaN,NaN,NaN,NaN,NaN,NaN,NaN,NaN,NaN,NaN,NaN,NaN,NaN,NaN,NaN,NaN
freq,565,NaN,NaN,NaN,NaN,NaN,NaN,NaN,NaN,NaN,NaN,NaN,NaN,NaN,NaN,NaN,NaN,NaN
mean,NaN,3001.638353,2018.804376,779.972973,27.558559,55.796654,3699.907336,855.298584,10440.669241,4357.526384,549.380952,1340.642214,72.660232,79.702703,14.089704,22.743887,9660.171171,65.46332
std,NaN,3870.201484,2451.113971,929.176190,17.640364,19.804778,4850.420531,1522.431887,4023.016484,1096.696416,165.105360,677.071454,16.328155,14.722359,3.958349,12.391801,5221.768440,17.17771
min,NaN,81.000000,72.000000,35.000000,1.000000,9.000000,139.000000,1.000000,2340.000000,1780.000000,96.000000,250.000000,8.000000,24.000000,2.500000,0.000000,3186.000000,10.00000
25%,NaN,776.000000,604.000000,242.000000,15.000000,41.000000,992.000000,95.000000,7320.000000,3597.000000,470.000000,850.000000,62.000000,71.000000,11.500000,13.000000,6751.000000,53.00000
50%,NaN,1558.000000,1110.000000,434.000000,23.000000,54.000000,1707.000000,353.000000,9990.000000,4200.000000,500.000000,1200.000000,75.000000,82.000000,13.600000,21.000000,8377.000000,65.00000
75%,NaN,3624.000000,2424.000000,902.000000,35.000000,69.000000,4005.000000,967.000000,12925.000000,5050.000000,600.000000,1700.000000,85.000000,92.000000,16.500000,31.000000,10830.000000,78.00000


## b)
***

In [16]:
#Dividing data 50/50 into trainind and test sets
np.random.seed(123)
n = College.shape[0]
train_idx = np.random.choice(n, size = n // 2, replace = False)
train = College.iloc[train_idx]
test = College.drop(College.index[train_idx])

## c)
***

In [17]:
#Fitting linear regression model
model2 = smf.ols('Apps ~ Private + Accept', data = train)
m2 = model2.fit()
print(m2.summary())

                            OLS Regression Results                            
Dep. Variable:                   Apps   R-squared:                       0.878
Model:                            OLS   Adj. R-squared:                  0.877
Method:                 Least Squares   F-statistic:                     1384.
Date:                Mon, 18 May 2026   Prob (F-statistic):          1.60e-176
Time:                        15:30:13   Log-Likelihood:                -3324.2
No. Observations:                 388   AIC:                             6654.
Df Residuals:                     385   BIC:                             6666.
Df Model:                           2                                         
Covariance Type:            nonrobust                                         
                     coef    std err          t      P>|t|      [0.025      0.975]
----------------------------------------------------------------------------------
Intercept        232.0093    186.224      1.

## d)
***

In [18]:
#Predicting on training data
pred_train = m2.predict(train)

#Compute MSE
mse_train = ((train['Apps'] - pred_train) ** 2).mean()
print(round(mse_train, 4))

1619152.4143


## e)
***

In [19]:
#C with only Accept as predictor
model3 = smf.ols('Apps ~ Accept', data = train)
m3 = model3.fit()
print(m3.summary())

                            OLS Regression Results                            
Dep. Variable:                   Apps   R-squared:                       0.878
Model:                            OLS   Adj. R-squared:                  0.877
Method:                 Least Squares   F-statistic:                     2770.
Date:                Mon, 18 May 2026   Prob (F-statistic):          3.30e-178
Time:                        15:30:13   Log-Likelihood:                -3324.6
No. Observations:                 388   AIC:                             6653.
Df Residuals:                     386   BIC:                             6661.
Df Model:                           1                                         
Covariance Type:            nonrobust                                         
                 coef    std err          t      P>|t|      [0.025      0.975]
------------------------------------------------------------------------------
Intercept    101.8338     85.355      1.193      0.2

In [20]:
#MSE for M3
pred_train_m3 = m3.predict(train)
mse_train_m3 = ((train['Apps'] - pred_train_m3) ** 2).mean()
print(round(mse_train_m3, 4))

1621754.6516


In [21]:
#RMSE for M2 and M3
rmse_m2 = np.sqrt(mse_train)
rmse_m3 = np.sqrt(mse_train_m3)
print(f'RMSE for M2: {rmse_m2:.4f}')
print(f'RMSE for M3: {rmse_m3:.4f}')

RMSE for M2: 1272.4592
RMSE for M3: 1273.4813


We clearly see that MSE for both `Accept` and `Private` yields the lower MSE, giving a RMSE of 1 272 against M3's RMSE of 1 273.

## f)
***

In [22]:
#Test-MSE for M2 and M3
pred_test_m2 = m2.predict(test)
mse_test_m2 = ((test['Apps'] - pred_test_m2) ** 2).mean()
print(f'Test MSE for M2: {mse_test_m2:.4f}')

pred_test_m3 = m3.predict(test)
mse_test_m3 = ((test['Apps'] - pred_test_m3) ** 2).mean()
print(f'Test MSE for M3: {mse_test_m3:.4f}')

Test MSE for M2: 1752114.9148
Test MSE for M3: 1726283.6547


As we can see, test MSE for M3 is lower than that of M2, suggesting that the predictor `Private` adds more noise than signal — the simpler model generalises better to new data. For both models, test-MSE is higher than training-MSE, as the models are optimised on training data and will always fit it better than unseen data.

## g)
***

In [23]:
#Experimenting with different random seeds
for seed in [1, 42, 99, 200, 999]:
    np.random.seed(seed)
    n = College.shape[0]
    train_idx = np.random.choice(n, size = n // 2, replace = False)
    train = College.iloc[train_idx]
    test = College.drop(College.index[train_idx])

    m2 = smf.ols('Apps ~ Private + Accept', data = train).fit()
    m3 = smf.ols('Apps ~ Accept', data = train).fit()

    mse_test_m2 = ((test['Apps'] - m2.predict(test)) ** 2).mean()
    mse_test_m3 = ((test['Apps'] - m3.predict(test)) ** 2).mean()

    print(f'Seed {seed}: M2 test-MSE = {mse_test_m2:.0f},  M3 test-MSE = {mse_test_m3:.0f}')


Seed 1: M2 test-MSE = 2039361,  M3 test-MSE = 2046138
Seed 42: M2 test-MSE = 1254612,  M3 test-MSE = 1223660
Seed 99: M2 test-MSE = 1763337,  M3 test-MSE = 1734008
Seed 200: M2 test-MSE = 1459877,  M3 test-MSE = 1455554
Seed 999: M2 test-MSE = 2057121,  M3 test-MSE = 2024883


## h)
***

In [24]:
#KNN for Apps
from sklearn.neighbors import KNeighborsRegressor
from sklearn.preprocessing import StandardScaler

In [25]:
X_train = train[['Accept']].values
X_test = test[['Accept']].values
y_train = train['Apps'].values
y_test = test['Apps'].values

In [26]:
#Standardising the predictor
scaler = StandardScaler()
X_train_scaled = scaler.fit_transform(X_train)
X_test_scaled = scaler.transform(X_test)

In [27]:
#KNN with multiple K values
for k in [1, 5, 10, 20, 50]:
    knn_model = KNeighborsRegressor(n_neighbors = k)
    knn_model.fit(X_train_scaled, y_train)
    pred_knn = knn_model.predict(X_test_scaled)
    mse_knn = ((y_test - pred_knn) ** 2).mean()
    print(f'KNN with K = {k}: Test MSE = {mse_knn:.0f}')

KNN with K = 1: Test MSE = 2778854
KNN with K = 5: Test MSE = 2083323
KNN with K = 10: Test MSE = 2130324
KNN with K = 20: Test MSE = 2304031
KNN with K = 50: Test MSE = 3429548


KNN performs worse than linear regression, as it misses heavily for universities with high application rates but similar acceptance rates. Linear regression follows the estimated line and handles these extreme values better. This is further highlighted by the $R^2$ of 0.878, indicating a clear linear relationship between Accept and Apps.